Improve emulated_hue compatibility with newer systems

Tho85 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

Add some missing fields and API routes to `emulated_hue` to make it discoverable by a Busch-Jaeger free@home SysAP.

free@home is a smart home solution distributed by Busch-Jaeger that consists of in-wall sensors/actors and a system access point (SysAP). The SysAP can discover Philips Hue bridges and control announced Hue bulbs.

This PR makes the following changes to the `emulated_hue` API:

* Send a correct serial number in `description.xml` and UPnP discovery packets. This seems to be the correct way to do it, since some devices try to derive a MAC address from the serial.
* Send a newer IPbridge version (1.16.0) as well as API version(1.17.0)
* Announce that the emulated bridge has a link button
* Add a `/api/{username}/config` route that returns the `config` part of the `/api/{username}` full-state route

The integration also continues to work at least with Amazon Alexa. Other devices were not tested due to lack of hardware.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Tho85 committed:
Make emulated hue detectable by Busch-Jaeger free@home SysAP

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -226,14 +226,47 @@ def get(self, request, username):
+            "swversion": "01003542",
+            "apiversion": "1.17.0",
+            "whitelist": {HUE_API_USERNAME: {"name": "HASS BRIDGE"}},
+            "ipaddress": f"{self.config.advertise_ip}:{self.config.advertise_port}",
```
I'm not sure this will work since `CONF_ADVERTISE_IP ` and `CONF_ADVERTISE_PORT` are optional.

bdraco replied:
```
        # Calculated effective advertised IP and port for network isolation
        self.advertise_ip = conf.get(CONF_ADVERTISE_IP) or self.host_ip_addr

        self.advertise_port = conf.get(CONF_ADVERTISE_PORT) or self.listen_port


```

Looks like we always set it so should be fine

bdraco commented:
<img width="890" alt="Screen Shot 2020-05-03 at 3 12 53 PM" src="https://user-images.githubusercontent.com/663432/80924789-9b434a80-8d50-11ea-9502-b05ce6307803.png">
Would you please add a test to make sure the endpoint isn't accessible from a public ip address

bdraco commented on homeassistant/components/emulated_hue/upnp.py:
```diff
@@ -74,13 +74,14 @@ def __init__(
         # this string is required per the SSDP spec
         resp_template = f"""HTTP/1.1 200 OK
-CACHE-CONTROL: max-age=60
+HOST: 239.255.255.250:1900
```
```suggestion
```

The host line doesn't look right

Tho85 replied:
I've removed this line, could you try again with your harmony hub?

As for the other changes in the upnp discovery response: The free@home sysap only works with the changes I made. It specifically looks for an `upnp:rootdevice`, so I hope that is not causing the issue with your harmony hub:

![image](https://user-images.githubusercontent.com/22068/80926299-b042cc80-8d96-11ea-8703-d24a439deee1.png)

bdraco commented:
Tested the changes with a harmony hub.  It wasn't able to find the bridge anymore

Tho85 committed:
Emulated hue: Remove unnecessary host line from UPnP response

Tho85 committed:
Test that external IPs are blocked for config route

bdraco commented:
Appears to be working now that the `Host` line was removed. 

Doing some more testing with it

Tho85 commented:
It is also discovered by the official "Philips Hue Bridge v1" Android app. (...which then complained about an outdated firmware on the "bridge"; the following firmware upgrade miraculously failed :wink: )

Tho85 commented:
... after changing the HUE_API_USERNAME to `nouser`, I forgot to add

bdraco commented:
> ... after changing the HUE_API_USERNAME to `nouser`, I forgot to add


Any downside you can think of to setting it to `nouser` in this PR?  Seems the closer it can get to the `hue` implementation the more compatible its likely to be.

Tho85 commented:
I was thinking about that as well. I'll test it with my devices and see what happens

Tho85 committed:
Add another test for unauthorized users

Tho85 committed:
Change hue username to nouser

nouser seems to be used by the official Hue Bridge v1 Android app and is
used by other projects as well

Tho85 commented:
`nouser` works with the original app, my Amazon Echo, and the free@home SysAP. It is also used by various other projects around the net, so let's stick with that.

bdraco commented:
@Tho85 Would you please fix the conflict with the other branch that just merged.  I'll test after dinner with `harmony`

bdraco committed:
Merge branch 'dev' into emulated-hue-freeathome

bdraco commented:
Working with harmony ok!

I'm also able to find the bridge with `Colors for Hue` mac desktop app.
